experimental/list_formatter/src/provider.rs

```diff
@@ -218,28 +225,29 @@ impl<'data> From<ListJoinerPattern<'data>> for ConditionalListJoinerPattern<'dat
 }
 
 impl<'a> ConditionalListJoinerPattern<'a> {
-    pub fn parts(&'a self, following_value: &str) -> PatternParts<'a> {
+    /// Returns the pattern parts, and if materialized, the following value as a string
+    pub fn parts<'b, W: FormattedWriteable + ?Sized>(
+        &'a self,
+        following_value: &'b W,
+    ) -> PatternParts<'a> {
         match &self.special_case {
-            Some(SpecialCasePattern { condition, pattern }) if condition.test(following_value) => {
+            Some(SpecialCasePattern { condition, pattern })
+                // TODO: Implement lookahead instead of materializing here.
+                if condition.test(following_value.writeable_to_fmt_string().as_str()) =>
```

Comment:
Suggestion: the new `writeable_to_string` impl that returns a `Cow<str>` is more efficient here and I think you should use it.  We can probably use this solution in the medium term.  We can still consider this a potential improvement in the future.

Reply:
`following_value` is a `FormattedWriteable`, and the `to_writeable` conversion only wraps that, so its `writeable_to_string` goes through the `FormattedWriteable` methods and doesn't support returning `Cow<str>`.

It would be nice to implement an efficient `to_writeable` on the blanket impl of `FormattedWriteable for W: Writeable`, but `to_writeable`'s type signature is too rigid to return `W` there...

Reply:
Hmm, this may be a good motivation to make `trait FormattedWriteable: Writeable`, and then everything that implements `FormattedWriteable` should also implement `Writeable` separately.